File: 2331. Evaluate Boolean Binary Tree/2331. Evaluate Boolean Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool EvaluateTree(TreeNode root) {
        if (root.val == 0)
            return false;
        else if (root.val == 1)
            return true;
        else if (root.val == 2)
            return EvaluateTree(root.left) || EvaluateTree(root.right);
        else
            return EvaluateTree(root.left) && EvaluateTree(root.right);
    }
}

File: 2331. Evaluate Boolean Binary Tree/2331. Evaluate Boolean Binary Tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn evaluate_tree(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if let Some(node) = root {
            let node = node.borrow();
            match node.val {
                0 => false,
                1 => true,
                2 => {
                    Solution::evaluate_tree(node.left.clone())
                        || Solution::evaluate_tree(node.right.clone())
                }
                3 => {
                    Solution::evaluate_tree(node.left.clone())
                        && Solution::evaluate_tree(node.right.clone())
                }
                _ => unreachable!(),
            }
        } else {
            unreachable!()
        }
    }
}